pkg/admission/mutator/shoot_test.go

```diff
@@ -17,70 +17,368 @@ package mutator
+				),
+
+				alicloudClientFactory.EXPECT().NewECSClient(regionId, accessKeyID, accessKeySecret).Return(ecsClient, nil),
+				ecsClient.EXPECT().CheckIfImageExists(ctx, imageId).Return(false, nil),
+				//ecsClient.EXPECT().CheckIfImageOwnedByAliCloud(imageId).Return(false, nil)
+			)
+			err := shoot.Mutate(ctx, newShoot, nil)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(*newShoot.Spec.Provider.Workers[0].Volume.Encrypted).To(BeTrue())
+			Expect(*newShoot.Spec.Provider.Workers[0].DataVolumes[0].Encrypted).To(BeTrue())
+			Expect(controllerutils.HasTask(newShoot.Annotations, v1beta1constants.ShootTaskDeployInfrastructure)).To(BeFalse())
+		})
+		It("should set encrypted flag as false for system disk if image is owned by alicloud", func() {
+			gomock.InOrder(
+				c.EXPECT().Get(ctx, kutil.Key("alicloud"), gomock.AssignableToTypeOf(&corev1beta1.CloudProfile{})).DoAndReturn(
+					func(_ context.Context, _ client.ObjectKey, obj *corev1beta1.CloudProfile) error {
+						*obj = *cloudProfile
+						return nil
+					},
+				),
+				c.EXPECT().Get(ctx, kutil.Key(namespace, name), gomock.AssignableToTypeOf(&corev1beta1.SecretBinding{})).DoAndReturn(
+					func(_ context.Context, _ client.ObjectKey, obj *corev1beta1.SecretBinding) error {
+						*obj = *secretBinding
+						return nil
+					},
+				),
+				apiReader.EXPECT().Get(ctx, kutil.Key(namespace, name), gomock.AssignableToTypeOf(&corev1.Secret{})).DoAndReturn(
+					func(_ context.Context, _ client.ObjectKey, obj *corev1.Secret) error {
+						*obj = *secret
+						return nil
+					},
+				),
+
+				alicloudClientFactory.EXPECT().NewECSClient(regionId, accessKeyID, accessKeySecret).Return(ecsClient, nil),
+				ecsClient.EXPECT().CheckIfImageExists(ctx, imageId).Return(true, nil),
+				ecsClient.EXPECT().CheckIfImageOwnedByAliCloud(imageId).Return(true, nil),
+			)
+			err := shoot.Mutate(ctx, newShoot, nil)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(newShoot.Spec.Provider.Workers[0].Volume.Encrypted == nil).To(BeTrue())
+			Expect(*newShoot.Spec.Provider.Workers[0].DataVolumes[0].Encrypted).To(BeTrue())
+		})
+		It("should set encrypted flag as true for newly added worker or datavolume", func() {
+			sameName := "worker1"
+			newName := "newWorker"
+
+			oldShoot.Spec.Provider.Workers[0].Volume.Encrypted = nil
+			newShoot.Spec.Provider.Workers[0].Volume.Encrypted = nil
+			oldShoot.Spec.Provider.Workers[0].Name = sameName
+			newShoot.Spec.Provider.Workers[0].Name = sameName
+
+			newShoot.Spec.Provider.Workers[1].Name = newName
+			newShoot.Spec.Provider.Workers[1].Volume = &corev1beta1.Volume{}
+			newShoot.Spec.Provider.Workers[1].DataVolumes = []corev1beta1.DataVolume{{}}
+			//simulate to add datavolume disk
+			oldShoot.Spec.Provider.Workers[0].DataVolumes[0].Name = "old"
+			newShoot.Spec.Provider.Workers[0].DataVolumes[0].Name = newName
+			oldShoot.Spec.Provider.Workers[0].DataVolumes[0].Encrypted = nil
+			newShoot.Spec.Provider.Workers[0].DataVolumes[0].Encrypted = nil
```

Comment:
Since a lot of fields are initialized, consider defining a utility function or initialize them like below:
```
oldShoot.Spec.Provider.Workers = []Workers{
  {
    Volume: xxx
    Name: xxx
    DataVolumes: []
  }
}
```

Reply:
Hi Emoin,
For different test cases, I mainly manipulate the name of system volume&data volume and encrypted flag. I went through it again, and found it can't make code simpler even if I add a utility function